project4/binomialengine.hpp

```diff
@@ -128,8 +129,26 @@ namespace QuantLib {
         // binomial tree 
         // (see J.C.Hull, "Options, Futures and other derivatives", 6th edition, pp 397/398)
 
+
         // Rollback to third-last step, and get underlying prices (s2) &
         // option values (p2) at this point
+
+
+        if (oscillations_ == true) {
+            option.rollback(grid[timeSteps_ - 1]);
+            Array vapenu(option.values());
+            QL_ENSURE(vapenu.size() == timeSteps_, "Expect time steps minus 1 nodes in grid at penultimate step");
+            Time dT = maturity / timeSteps_;
+            for (int i = 0; i < timeSteps_; i++) {
+                Real u_pen = lattice->underlying(timeSteps_ - 1, i);
+                BlackCalculator bs_option(payoff->optionType(), payoff->strike(), u_pen * std::exp((r - q) * dT), v*std::sqrt(dT), std::exp(-r*dT));
+                option.values()[i] = bs_option.value();
```

Comment:
One last thing: it doesn't change the values significantly, but this should be
```suggestion
                option.values()[i] = std::max(bs_option.value(), (*payoff)(u_pen));
```
because the holder of the option can exercise it and get the payoff if it's worth more then the option value.